File: csharp/ivr7helper.cs
using System.Collections.Generic;
using System;
class ivr7helper
{
  public static void makeDistanceTimeCapDims(Ivr7.Model m)
  {
    m.Dimensions = new Ivr7.DimensionConfiguration();

    m.Dimensions.timeConfig = new Ivr7.InternalDimension
    {
      Id = "time",
      measurementUnit = Ivr7.InternalDimension.eMeasurementUnit.Minutes,
      slackMax = (float)1e6,
      tardyMax = 0
    };
    m.Dimensions.distanceConfig = new Ivr7.InternalDimension
    {
      Id = "distance",
      measurementUnit = Ivr7.InternalDimension.eMeasurementUnit.Kilometres,
      slackMax = 0,
      tardyMax = 0
    };

    m.Dimensions.capacityDimensions.Add(
      new Ivr7.CapacityDimension
      {
        Id = "capacity",
        Units = "kg",
        slackMax = 0,
        tardyMax = 0
      }
    );
  }

  public static void makeLocations(Ivr7.Model m, List<dataRow> d)
  {
    m.Locations.Clear();
    foreach (var l in d)
    {
      m.Locations.Add(new Ivr7.Location
      {
        Id = l.id,
        Geocode = new Ivr7.Geocode
        {
          Longitude = l.X,
          Latitude = l.Y
        }
      });

    }
  }

  public static void makeJobTimeCap(Ivr7.Model m, List<dataRow> d, List<int> srcs, List<int> dests)
  {
    m.Jobs.Clear();
    if (srcs.Count != dests.Count)
    {
      throw new System.Exception("Expected srcs.Count == dests.Count");
    }
    for (int i = 0; i < srcs.Count; i++)
    {
      int si = srcs[i];
      int di = dests[i];
      Ivr7.Job j = new Ivr7.Job();
      j.Id = "job_" + d[di].id;
      j.pickupTask = new Ivr7.Job.Task();
      j.pickupTask.taskId = "pickup_" + d[di].id;
      j.pickupTask.locationId = d[si].id;

      j.pickupTask.Attributes.Add(
        new Ivr7.Job.Task.Attribute
        {
          dimensionId = "time",
          Quantity = d[di].pickupTime
        });

      j.pickupTask.Attributes.Add(
        new Ivr7.Job.Task.Attribute
        {
          dimensionId = "capacity",
          Quantity = d[di].quanity
        }
      );

      j.dropoffTask = new Ivr7.Job.Task();
      j.dropoffTask.taskId = "dropoff_" + d[di].id;
      j.dropoffTask.locationId = d[di].id;
      // careful here, in C# the object is passed by ref.
      // so instantiate a new set of attributes, don't reuse the object's created
      // prior to this.
      j.dropoffTask.Attributes.Add(
        new Ivr7.Job.Task.Attribute
        {
          dimensionId = "time",
          Quantity = d[di].dropoffTime
        });

      j.dropoffTask.Attributes.Add(
        new Ivr7.Job.Task.Attribute
        {
          dimensionId = "capacity",
          Quantity = -d[di].quanity
        }
      );
      j.Penalty = 10000;
      m.Jobs.Add(j);
    }
  }

  public static Ivr7.VehicleClass makeVcSimple(string name, float time_transit_coef,
                                               float time_task_coef, float time_loc_coef, float distance_transit_coef)
  {
    Ivr7.VehicleClass vc = new Ivr7.VehicleClass
    {
      Id = name
    };
    vc.Attributes.Add(
      new Ivr7.VehicleClass.Attribute
      {
        dimensionId = "time",
        transitCoef = time_transit_coef,
        taskCoef = time_task_coef,
        locationCoef = time_loc_coef,
        transitGeneratorId = "roadnetwork_time"
      });
    vc.Attributes.Add(
      new Ivr7.VehicleClass.Attribute
      {
        dimensionId = "distance",
        transitCoef = distance_transit_coef,
        transitGeneratorId = "roadnetwork_distance"
      });
    return vc;
  }

  public static Ivr7.VehicleCostClass makeVccSimple(string name, float fixedcost, float time_transit_costcoef,
                          float time_loc_costcoef, float time_task_costcoef, float time_slack_costcoef,
                          float distance_transit_costcoef)
  {
    Ivr7.VehicleCostClass vcc = new Ivr7.VehicleCostClass
    {
      Id = name,
      fixedCost = fixedcost
    };
    vcc.Attributes.Add(new Ivr7.VehicleCostClass.Attribute
    {
      dimensionId = "time",
      transitCostCoef = time_transit_costcoef,
      taskCostCoef = time_task_costcoef,
      locationCostCoef = time_loc_costcoef,
      slackCostCoef = time_slack_costcoef
    });

    vcc.Attributes.Add(new Ivr7.VehicleCostClass.Attribute
    {
      dimensionId = "distance",
      transitCostCoef = distance_transit_costcoef
    });
    return vcc;
  }

  public static Ivr7.Vehicle makeVehicleCap(string name, string vehicleClass, string vehicleCostClass, float capacity,
  string startLocation, string endLocation, float startTime, float endTime)
  {
    Ivr7.Vehicle v = new Ivr7.Vehicle
    {
      Id = name,
      classId = vehicleClass,
      costClassId = vehicleCostClass,
      shift = new Ivr7.Vehicle.Shift
      {
        shiftStart = new Ivr7.Vehicle.Task
        {
          locationId = startLocation,
        },
        shiftEnd = new Ivr7.Vehicle.Task
        {
          locationId = endLocation
        }
      }
    };
    // then just add the time windows to the shift
    v.shift.shiftStart.Attributes.Add(new Ivr7.Vehicle.Task.Attribute
    {
      dimensionId = "time",
      Quantity = 0,
      Windows = { new Ivr7.Window { Start = startTime, End = endTime } }
    });

    v.shift.shiftEnd.Attributes.Add(new Ivr7.Vehicle.Task.Attribute
    {
      dimensionId = "time",
      Quantity = 0,
      Windows = { new Ivr7.Window { Start = startTime, End = endTime } }
    });
    // lastly, set the capacity of the vehicle.
    v.Capacities.Add(new Ivr7.Vehicle.Capacity
    {
      dimensionId = "capacity",
      capacity = capacity
    });
    return v;
  }

  public static Ivr7.TransitRule makeLunchBreakRule(string name, string prefix, float lunchtime, float breaklength)
  {
    var tr = new Ivr7.TransitRule
    {
      Id = name,
      ruleIdPrefix = prefix,
      dimensionId = "time",
      useStandingState = true,
      useTransitState = true,
      Triggers = {
            new Ivr7.TransitRule.Trigger{
              Value = lunchtime,
              Quantity = breaklength
            }
          }
    };
    return tr;
  }


  class agg
  {
    public void Add(float delta, float cost)
    {
      cumul += delta;
      min = Math.Min(cumul, min);
      max = Math.Max(cumul, max);
      totalCost += cost;
    }
    public float min = (float)1e30;
    public float max = 0;
    public float cumul = 0;
    public float totalCost = 0;
  };

  public static void printSolution(Ivr7.SolutionResponse solution, bool printnodes = true,
                                                                  bool printaggstats = true,
                                                                  bool printTrrules = true,
                                                                  bool printInfeas = true)
  {

    int[] maxchar = new int[5] {"stopId".Length,
                                "locationId".Length,
                                "taskId".Length,
                                "jobId".Length,
                                "vehicleId".Length }; // column widths for Stopid locationId, taskId, jobId, vehicleId
    foreach (var r in solution.Routes)
    {
      maxchar[3] = Math.Max(maxchar[3], r.vehicleId.Length);
      foreach (var s in r.Stops)
      {
        maxchar[0] = Math.Max(maxchar[0], s.locationId.Length);
        maxchar[1] = Math.Max(maxchar[1], s.taskId.Length);
        maxchar[2] = Math.Max(maxchar[2], s.jobId.Length);
      }
    }
    string formatLine = "";
    if (printnodes)
    {
      formatLine = "|{0,-7}|{1,-" + maxchar[0] + "}|{2,-" + maxchar[1] + "}|{3,-" + maxchar[2] + "}|{4,-" + maxchar[3] + "}|";
      Console.WriteLine(String.Format(formatLine, "stopId", "locationId", "taskId", "jobId", "vehicleId"));
      foreach (var r in solution.Routes)
      {
        if (r.Stops.Count > 2)
        {
          foreach (var s in r.Stops)
          {
            Console.WriteLine(String.Format(formatLine, s.Id, s.locationId, s.taskId, s.jobId, r.vehicleId));
          }
          Console.WriteLine(""); // pop a gap in for the next vehicle.
        }
      }
    }
    // find all the dimensions in the model
    HashSet<string> dims = new HashSet<string>();
    foreach (var r in solution.Routes)
    {
      foreach (var s in r.Stops)
      {
        foreach (var a in s.Attributes)
        {
          dims.Add(a.dimId);
        }
      }
    }

    // because tables can't be too wide in c# we're just going to summarise the solution here by vehicle
    // into the cumulative, min, and max quantities per vehicle/dimension.
    Dictionary<string, Dictionary<string, agg>> vehicleStats = new Dictionary<string, Dictionary<string, agg>>();

    int dimLenChar = "dimension".Length;
    int ruleIdCharLen = "ruleId".Length;
    bool hasTrRules = false;

    foreach (var d in dims)
    {
      dimLenChar = Math.Max(dimLenChar, d.Length);
    }

    foreach (var r in solution.Routes)
    {
      vehicleStats.Add(r.vehicleId, new Dictionary<string, agg>());
      foreach (var d in dims)
      {
        vehicleStats[r.vehicleId].Add(d, new agg());
      }
      foreach (var s in r.Stops)
      {
        foreach (var a in s.Attributes)
        {
          vehicleStats[r.vehicleId][a.dimId].Add(a.endValue - a.startValue, a.Cost);
        }
      }
      foreach (var s in r.interStops)
      {
        foreach (var a in s.Attributes)
        {
          vehicleStats[r.vehicleId][a.dimId].Add(a.endValue - a.startValue, a.Cost);
        }
      }
      foreach (var tr in r.transitRuleAttributes)
      {
        ruleIdCharLen = Math.Max(ruleIdCharLen, tr.ruleId.Length);
        hasTrRules = true;
      }
    }
    // lets print the summary data out. |vehicleId|dimension|Min|Max|Cumul|Cost
    if (printaggstats)
    {
      Console.WriteLine("Vehilce dimension summary");
      formatLine = "|{0,-" + maxchar[3] + "}|{1,-" + dimLenChar + "}|{2,10}|{3,10}|{4,10}|{5,10}|";
      Console.WriteLine(String.Format(formatLine, "vehicleId", "dimension", "Min", "Max", "Cumul", "Cost"));
      foreach (var kvp in vehicleStats)
      {
        foreach (var l in kvp.Value)
        {
          if (l.Value.min != (float)(1e30))
          { // otherwise we know it had no values
            Console.WriteLine(String.Format(formatLine, kvp.Key, l.Key,
                              string.Format("{0:0.00}", l.Value.min),
                              string.Format("{0:0.00}", l.Value.max),
                              string.Format("{0:0.00}", l.Value.cumul),
                              string.Format("{0:0.00}", l.Value.totalCost)));
          }
        }
      }
    }
    // tabulate the tranist rules.
    // vehicleId|ruleId|dimID|fromstopid|tostopid|startvalue|endvalue|cost
    if (hasTrRules && printTrrules)
    {
      Console.WriteLine("\nVehilce transit rule summary:");
      formatLine = "|{0,-" + maxchar[3] + "}|{1," + ruleIdCharLen + "}|{2,-" + dimLenChar + "}|{3,10}|{4,10}|{5,10}|{6,10}|{7,10}|";
      Console.WriteLine(String.Format(formatLine, "vehicleId", "ruleId", "dimension", "fromStopId", "toStopId", "startValue", "endValue", "cost"));
      foreach (var r in solution.Routes)
      {
        foreach (var tr in r.transitRuleAttributes)
        {
          Console.WriteLine(String.Format(formatLine,
                            r.vehicleId, tr.ruleId, tr.dimId, tr.fromStopId, tr.toStopId,
                            string.Format("{0:0.00}", tr.startValue),
                            string.Format("{0:0.00}", tr.endValue),
                            string.Format("{0:0.00}", tr.Cost)));
        }
      }
    }
    // lastly, lets tabulate the infeasibilities
    if (printInfeas)
    {
      Console.WriteLine("\nInfeasibilities:");
      //"dimension|message|limit|value|count|taskId"
      int messageCharLen = "message".Length;
      foreach (var i in solution.Infeasibilities)
      {
        foreach (var m in i.infeasibilityInfoes)
        {
          messageCharLen = Math.Max(messageCharLen, m.Message.Length);
          dimLenChar = Math.Max(m.dimId.Length, dimLenChar);
        }
      }
      formatLine = "|{0,-" + dimLenChar + "}|{1,-" + messageCharLen + "}|{2,10}|{3,10}|{4,10}|{5," + maxchar[1] + "}|";
      Console.WriteLine(String.Format(formatLine, "dimenion", "message", "limit", "value", "count", "taskId"));
      foreach (var i in solution.Infeasibilities)
      {
        foreach (var m in i.infeasibilityInfoes)
        {
          Console.WriteLine(String.Format(formatLine,
                        m.dimId, m.Message,
                        string.Format("{0:0.00}", m.Limit),
                        string.Format("{0:0.00}", m.Value),
                        string.Format("{0:0.00}", m.Count),
                        i.taskId));
        }
      }
    }

  }

  // Repeat a value to a certain count
  public static List<int> Rep(int value, int count)
  {
    List<int> res = new List<int>(count);
    for (int i = 0; i < count; i++)
    {
      res.Add(value);
    }
    return res;
  }

  // Create a sequence [from:to) (exclusive of the to)
  // assumes you're not going to call this with [-int.MinValue, int.MaxValue] :-p
  public static List<int> Seq(int from, int to)
  {
    List<int> res = new List<int>();
    while (from != to)
    {
      res.Add(from);
      from++;
    }
    return res;
  }

}

File: java/src/main/java/icepackai/ivr7helper.java
package icepackai;

import java.util.*;

import icepackai.IVR7.Ivr7Kt461V8Eoaif;
import icepackai.IVR7.Ivr7Kt461V8Eoaif.InternalDimension.eMeasurementUnit;

import dnl.utils.text.table.TextTable;

// some helper functions for building ivr7 models.
public class ivr7helper {
  public static void makeDistanceTimeCapDims(Ivr7Kt461V8Eoaif.Model.Builder m) {
    Ivr7Kt461V8Eoaif.DimensionConfiguration.Builder dimBuilder =
        Ivr7Kt461V8Eoaif.DimensionConfiguration.newBuilder();

    dimBuilder.setTimeConfig(Ivr7Kt461V8Eoaif.InternalDimension.newBuilder()
                                 .setId("time")
                                 .setMeasurementUnit(eMeasurementUnit.MINUTES)
                                 .setSlackMax(1e6f)
                                 .setTardyMax(0)
                                 .build());

    dimBuilder.setDistanceConfig(Ivr7Kt461V8Eoaif.InternalDimension.newBuilder()
                                     .setId("distance")
                                     .setMeasurementUnit(eMeasurementUnit.KILOMETRES)
                                     .setSlackMax(0)
                                     .setTardyMax(0)
                                     .build());

    dimBuilder.addCapacityDimensions(Ivr7Kt461V8Eoaif.CapacityDimension.newBuilder()
                                         .setId("capacity")
                                         .setUnits("kg")
                                         .setSlackMax(0)
                                         .setTardyMax(0)
                                         .build());

    m.setDimensions(dimBuilder.build());
  }

  public static void makeLocations(
      Ivr7Kt461V8Eoaif.Model.Builder m, List<dataRow> d, float windowStart, float windowEnd) {
    for (int i = 0; i < d.size(); i++) {
      dataRow loc = d.get(i);
      if (windowEnd == 0.0f && windowStart == 0.0f) {
        m.addLocations(
            Ivr7Kt461V8Eoaif.Location.newBuilder()
                .setId(loc.id)
                .setGeocode(
                    Ivr7Kt461V8Eoaif.Geocode.newBuilder().setLatitude(loc.Y).setLongitude(loc.X))
                .build());
      } else {
        m.addLocations(
            Ivr7Kt461V8Eoaif.Location.newBuilder()
                .setId(loc.id)
                .setGeocode(
                    Ivr7Kt461V8Eoaif.Geocode.newBuilder().setLatitude(loc.Y).setLongitude(loc.X))
                .addAttributes(
                    Ivr7Kt461V8Eoaif.Location.Attribute.newBuilder()
                        .setDimensionId("time")
                        .addArrivalWindows(
                            Ivr7Kt461V8Eoaif.Window.newBuilder().setStart(8 * 60f).setEnd(
                                14 * 60f)))
                .build());
      }
    }
  }

  public static void makeLocations(Ivr7Kt461V8Eoaif.Model.Builder m, List<dataRow> d) {
    makeLocations(m, d, 0.0f, 0.0f);
  }

  public static void makeJobTimeCap(Ivr7Kt461V8Eoaif.Model.Builder m, List<dataRow> d,
      List<Integer> srcs, List<Integer> dests) throws Exception {
    if (srcs.size() != dests.size()) {
      throw new Exception("Expected srcs.Count == dests.Count");
    }
    for (int i = 0; i < srcs.size(); i++) {
      Integer si = srcs.get(i);
      Integer di = dests.get(i);
      Ivr7Kt461V8Eoaif.Job.Builder job = Ivr7Kt461V8Eoaif.Job.newBuilder();
      job.setId("job_" + d.get(di).id);
      job.setPickupTask(Ivr7Kt461V8Eoaif.Job.Task.newBuilder()
                            .setTaskId("pickup_" + d.get(di).id)
                            .setLocationId(d.get(si).id)
                            .addAttributes(Ivr7Kt461V8Eoaif.Job.Task.Attribute.newBuilder()
                                               .setDimensionId("time")
                                               .setQuantity(d.get(di).pickupTime)
                                               .build())
                            .addAttributes(Ivr7Kt461V8Eoaif.Job.Task.Attribute.newBuilder()
                                               .setDimensionId("capacity")
                                               .setQuantity(d.get(di).quantity)
                                               .build())
                            .build());

      job.setDropoffTask(Ivr7Kt461V8Eoaif.Job.Task.newBuilder()
                             .setTaskId("dropoff_" + d.get(di).id)
                             .setLocationId(d.get(di).id)
                             .addAttributes(Ivr7Kt461V8Eoaif.Job.Task.Attribute.newBuilder()
                                                .setDimensionId("time")
                                                .setQuantity(d.get(di).dropoffTime)
                                                .build())
                             .addAttributes(Ivr7Kt461V8Eoaif.Job.Task.Attribute.newBuilder()
                                                .setDimensionId("capacity")
                                                .setQuantity(-d.get(di).quantity)
                                                .build())
                             .build());
      job.setPenalty(10000f);
      m.addJobs(job.build());
    }
  }

  public static Ivr7Kt461V8Eoaif.VehicleCostClass makeVccSimple(String name, float fixedcost,
      float time_transit_costcoef, float time_loc_costcoef, float time_task_costcoef,
      float time_slack_costcoef, float distance_transit_costcoef) {
    return Ivr7Kt461V8Eoaif.VehicleCostClass.newBuilder()
        .setId(name)
        .setFixedCost(fixedcost)
        .addAttributes(Ivr7Kt461V8Eoaif.VehicleCostClass.Attribute.newBuilder()
                           .setDimensionId("time")
                           .setTransitCostCoef(time_transit_costcoef)
                           .setTaskCostCoef(time_task_costcoef)
                           .setLocationCostCoef(time_loc_costcoef)
                           .setSlackCostCoef(time_slack_costcoef)
                           .build())
        .addAttributes(Ivr7Kt461V8Eoaif.VehicleCostClass.Attribute.newBuilder()
                           .setDimensionId("distance")
                           .setTransitCostCoef(distance_transit_costcoef)
                           .build())
        .build();
  }

  public static Ivr7Kt461V8Eoaif.VehicleClass makeVcSimple(String name, float time_transit_coef,
      float time_task_coef, float time_loc_coef, float distance_transit_coef) {
    return Ivr7Kt461V8Eoaif.VehicleClass.newBuilder()
        .setId(name)
        .addAttributes(Ivr7Kt461V8Eoaif.VehicleClass.Attribute.newBuilder()
                           .setDimensionId("time")
                           .setTransitGeneratorId("roadnetwork_time")
                           .setTransitCoef(time_transit_coef)
                           .setTaskCoef(time_task_coef)
                           .setLocationCoef(time_loc_coef)
                           .build())
        .addAttributes(Ivr7Kt461V8Eoaif.VehicleClass.Attribute.newBuilder()
                           .setDimensionId("distance")
                           .setTransitGeneratorId("roadnetwork_distance")
                           .setTransitCoef(distance_transit_coef)
                           .build())
        .build();
  }

  public static Ivr7Kt461V8Eoaif.Vehicle makeVehicleCap(String name, String vehicleClass,
      String vehicleCostClass, float capacity, String startLocation, String endLocation,
      float startTime, float endTime) {
    return Ivr7Kt461V8Eoaif.Vehicle.newBuilder()
        .setId(name)
        .setClassId(vehicleClass)
        .setCostClassId(vehicleCostClass)
        .setShift(
            Ivr7Kt461V8Eoaif.Vehicle.Shift.newBuilder()
                .setShiftStart(
                    Ivr7Kt461V8Eoaif.Vehicle.Task.newBuilder()
                        .setLocationId(startLocation)
                        .addAttributes(
                            Ivr7Kt461V8Eoaif.Vehicle.Task.Attribute.newBuilder()
                                .setDimensionId("time")
                                .setQuantity(0.0f)
                                .addWindows(
                                    Ivr7Kt461V8Eoaif.Window.newBuilder().setStart(startTime).setEnd(
                                        endTime))
                                .build())
                        .build())
                .setShiftEnd(Ivr7Kt461V8Eoaif.Vehicle.Task.newBuilder()
                                 .setLocationId(endLocation)
                                 .addAttributes(Ivr7Kt461V8Eoaif.Vehicle.Task.Attribute.newBuilder()
                                                    .setDimensionId("time")
                                                    // then just add the time windows to the shift
                                                    .setQuantity(0.0f)
                                                    .addWindows(Ivr7Kt461V8Eoaif.Window.newBuilder()
                                                                    .setStart(startTime)
                                                                    .setEnd(endTime))
                                                    .build())
                                 .build())
                .build())
        .addCapacities(Ivr7Kt461V8Eoaif.Vehicle.Capacity.newBuilder()
                           .setDimensionId("capacity")
                           .setCapacity(capacity)
                           .build())
        .build();
  }

  public static Ivr7Kt461V8Eoaif.TransitRule makeLunchBreakRule(
      String name, String prefix, float lunchtime, float breaklength) {
    return Ivr7Kt461V8Eoaif.TransitRule.newBuilder()
        .setId(name)
        .setDimensionId("time")
        .setRuleIdPrefix(prefix)
        .setUseStandingState(true)
        .setUseTransitState(true)
        .addTriggers(
            Ivr7Kt461V8Eoaif.TransitRule.Trigger.newBuilder().setValue(lunchtime).setQuantity(
                breaklength))
        .build();
  }

  public static void printSolution(Ivr7Kt461V8Eoaif.SolutionResponse solution, Boolean printnodes,
      Boolean printaggstats, Boolean printTrrules, Boolean printInfeas) {
    class agg {
      public void add(float delta, float cost) {
        cumul += delta;
        min = Math.min(cumul, min);
        max = Math.max(cumul, max);
        totalCost += cost;
      }

      public float min = (float) 1e30;
      public float max = 0;
      public float cumul = 0;
      public float totalCost = 0;
    };
    if (printnodes) {
      String[] columnNames = {"stopId", "locationId", "taskId", "jobId", "vehicleId"};
      List<Object[]> tabData = new ArrayList<Object[]>();

      for (int i = 0; i < solution.getRoutesCount(); i++) {
        Ivr7Kt461V8Eoaif.SolutionResponse.Route r = solution.getRoutes(i);
        if (r.getStopsCount() > 2) {
          for (int j = 1; j < r.getStopsCount() - 1; j++) {
            Ivr7Kt461V8Eoaif.SolutionResponse.Stop s = r.getStops(j);
            tabData.add(new Object[] {
                s.getId(), s.getLocationId(), s.getTaskId(), s.getJobId(), r.getVehicleId()});
          }
          if (i != solution.getRoutesCount() - 1) {
            tabData.add(new Object[] {"", "", "", "", ""});
          }
        }
      }
      printTable(tabData, columnNames);
    }
    HashSet<String> dims = new HashSet<String>();
    for (Ivr7Kt461V8Eoaif.SolutionResponse.Route r : solution.getRoutesList()) {
      for (Ivr7Kt461V8Eoaif.SolutionResponse.Stop s : r.getStopsList()) {
        for (Ivr7Kt461V8Eoaif.SolutionResponse.StopAttribute a : s.getAttributesList()) {
          dims.add(a.getDimId());
        }
      }
    }

    Boolean hasTrRules = false;
    Map<String, HashMap<String, agg>> vehicleStats = new HashMap<String, HashMap<String, agg>>();

    for (Ivr7Kt461V8Eoaif.SolutionResponse.Route r : solution.getRoutesList()) {
      vehicleStats.put(r.getVehicleId(), new HashMap<String, agg>());
      for (String dim : dims) {
        vehicleStats.get(r.getVehicleId()).put(dim, new agg());
      }
      hasTrRules = hasTrRules || r.getTransitRuleAttributesCount() > 0;

      for (Ivr7Kt461V8Eoaif.SolutionResponse.Stop s : r.getStopsList()) {
        for (Ivr7Kt461V8Eoaif.SolutionResponse.StopAttribute a : s.getAttributesList()) {
          vehicleStats.get(r.getVehicleId())
              .get(a.getDimId())
              .add(a.getEndValue() - a.getStartValue(), a.getCost());
        }
      }

      for (Ivr7Kt461V8Eoaif.SolutionResponse.InterStop s : r.getInterStopsList()) {
        for (Ivr7Kt461V8Eoaif.SolutionResponse.InterStopAttribute a : s.getAttributesList()) {
          vehicleStats.get(r.getVehicleId())
              .get(a.getDimId())
              .add(a.getEndValue() - a.getStartValue(), a.getCost());
        }
      }
    }

    if (printaggstats) {
      System.out.println("\nVehilce dimension summary");
      String[] columnNames = {"vehicleId", "dimension", "Min", "Max", "Cumul", "Cost"};
      List<Object[]> tabData = new ArrayList<Object[]>();
      for (Map.Entry<String, HashMap<String, agg>> vSet : vehicleStats.entrySet()) {
        for (Map.Entry<String, agg> stat : vSet.getValue().entrySet()) {
          agg a = stat.getValue();
          if (a.min != 1e30f) {
            tabData.add(
                new Object[] {vSet.getKey(), stat.getKey(), a.min, a.max, a.cumul, a.totalCost});
          }
        }
      }
      printTable(tabData, columnNames);
    }

    if (hasTrRules && printTrrules) {
      System.out.println("\nVehilce transit rule summary:");
      String[] columnNames = new String[] {"vehicleId", "ruleId", "dimension", "fromStopId",
          "toStopId", "startValue", "endValue", "cost"};
      List<Object[]> tabData = new ArrayList<Object[]>();
      for (Ivr7Kt461V8Eoaif.SolutionResponse.Route r : solution.getRoutesList()) {
        for (Ivr7Kt461V8Eoaif.SolutionResponse.TransitRuleAttribute tr :
            r.getTransitRuleAttributesList()) {
          tabData.add(
              new Object[] {r.getVehicleId(), tr.getRuleId(), tr.getDimId(), tr.getFromStopId(),
                  tr.getToStopId(), tr.getStartValue(), tr.getEndValue(), tr.getCost()});
        }
      }
      printTable(tabData, columnNames);
    }

    if (printInfeas) {
      System.out.println("\nInfeasibilities:");
      String[] columnNames =
          new String[] {"dimenion", "message", "limit", "value", "count", "taskId"};
      List<Object[]> tabData = new ArrayList<Object[]>();
      for (Ivr7Kt461V8Eoaif.SolutionResponse.Infeasibility i : solution.getInfeasibilitiesList()) {
        for (Ivr7Kt461V8Eoaif.SolutionResponse.Infeasibility.Info m :
            i.getInfeasibilityInfoList()) {
          tabData.add(new Object[] {m.getDimId(), m.getMessage(), m.getLimit(), m.getValue(),
              m.getCount(), i.getTaskId()});
        }
      }
      printTable(tabData, columnNames);
    }
  }

  private static void printTable(List<Object[]> tabData, String[] columnNames) {
    Object[][] tab = new Object[tabData.size()][];
    for (int i = 0; i < tabData.size(); i++) {
      tab[i] = tabData.get(i);
    }
    TextTable tt = new TextTable(columnNames, tab);
    tt.printTable();
  }

  // Repeat a value to a certain count
  public static List<Integer> Rep(Integer value, Integer count) {
    List<Integer> res = new ArrayList<Integer>();
    for (int i = 0; i < count; i++) {
      res.add(value);
    }
    return res;
  }

  // Create a sequence [from:to) (exclusive of the to)
  // assumes you're not going to call this with [-int.MinValue, int.MaxValue] :-p
  public static List<Integer> Seq(Integer from, Integer to) {
    List<Integer> res = new ArrayList<Integer>();
    while (from != to) {
      res.add(from);
      from++;
    }
    return res;
  }
}